docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md

```diff
@@ -0,0 +1,203 @@
+
+## Introduction
+Graphics shaders are programs to run on a Graphics Processing Unit. Shaders can be written using the *OpenGL Shading Language* or *SPIR-V* (a simple binary intermediate language). Since shaders are programs, they need to be compiled into specific GPU instructions. A question that arises is what if the shader compiler has bugs? Some tools have been developed to deal with this problem.
+
+[SPIRV-Tools](https://github.com/KhronosGroup/SPIRV-Tools) is a collection of tools to work on *SPIR-V* modules.
+One of these tools is *spirv-fuzz*. It applies semantics-preserving transformations to a *SPIR-V* module. Theoretically, a semantics-preserving sentence should not change the behaviour of the variant code. If it has changed, then an error has occurred halfway, probably a compiler bug. In this way, the reference and variant shaders can be compared to check the differences. This approach, known as fuzzing, is part of technique called metamorphic testing. 
+
+The following is an example of transformation that replaces an `OpVectorTimesScalar` instruction with its mathematical definition. The vector components are extracted and multiplied by the scalar. Finally, the new vector is constructed using the multiplied components.
+
+<table>
+  <tr>
+    <th>
+      Reference shader
+    </th>
+    <th>
+      Variant shader
+    </th>
+  </tr>
+  <tr>
+    <td>
+      <pre>
+        <code>
+      OpCapability Shader
+ %1 = OpExtInstImport "GLSL.std.450"
+      OpMemoryModel Logical GLSL450
+      OpEntryPoint Vertex %10 "main"
+ %2 = OpTypeFloat 32
+ %3 = OpTypeVoid
+ %4 = OpTypeFunction %3
+ %5 = OpTypeVector %2 2
+ %6 = OpConstant %2 1
+ %7 = OpConstant %2 2
+ %8 = OpConstant %2 3
+ %9 = OpConstantComposite %5 %6 %7
+%10 = OpFunction %3 None %4
+%11 = OpLabel
+<strong>%12 = OpVectorTimesScalar %5 %9 %8</strong>
+      OpReturn
+      OpFunctionEnd
+        </code>
+      </pre>
+    </td>
+    <td>
+      <pre>
+        <code>
+      OpCapability Shader
+ %1 = OpExtInstImport "GLSL.std.450"
+      OpMemoryModel Logical GLSL450
+      OpEntryPoint Vertex %10 "main"
+ %2 = OpTypeFloat 32
+ %3 = OpTypeVoid
+ %4 = OpTypeFunction %3
+ %5 = OpTypeVector %2 2
+ %6 = OpConstant %2 1
+ %7 = OpConstant %2 2
+ %8 = OpConstant %2 3
+ %9 = OpConstantComposite %5 %6 %7
+%10 = OpFunction %3 None %4
+%11 = OpLabel
+<strong>%13 = OpCompositeExtract %2 %9 0
+%14 = OpFMul %2 %13 %8
+%15 = OpCompositeExtract %2 %9 1
+%16 = OpFMul %2 %15 %8
+%12 = OpCompositeConstruct %5 %14 %16</strong>
+      OpReturn
+      OpFunctionEnd
+        </code>
+      </pre>
+    </td>
+  </tr>
+</table>
+
+This project was concerned to use the metamorphic testing approach to extend *spirv-fuzz* by adding new transformations to cover features of *SPIR-V* that are not yet very widely-used, but that will be in the future.
```

Comment:
I would also remove: "to cover features of *SPIR-V* that are not yet very widely-used, but that will be in the future"

I don't think this is true. We want to target features that are used and will continue to be used. You could say that you added transformations to cover instructions and patterns of SPIR-V that were not previously covered by existing spirv-fuzz transformations.

Reply:
As a counter-argument to what @paulthomson said: we have been keen to cover parts of SPIR-V that are not widely-used by GLSL shaders that get converted to SPIR-V by glslang.  I think some of the transformations that we make lead to SPIR-V that you could never get from one of the standard front-ends, so in that sense it's SPIR-V that isn't widely-used.  You might want to discuss that a bit Andre.